README.md
# Curvy connections

Library to draw connection arrow for graphs, flow charts, mind maps, etc.




CHANGELOG.md
# 1.0.2

+ Fixed curve when start.dx not 0
+ Fixed repaint

# 1.0.1

+ Converted `child` parameter in `ScrollView` to named parameter
+ Updated LICENSE

# 1.0.0

+ Draw curvy connections

lib/arrow_painter.dart
import 'dart:math';

import 'package:arrow_path/arrow_path.dart';
import 'package:flutter/material.dart';

class Connection {
  final Offset start;
  final Offset end;
  final double bend;
  final double tipAngle;
  final double tipLength;

  Connection(
      {required this.start,
      required this.end,
      this.bend = 0.5,
      this.tipAngle = 30,
      this.tipLength = 15});

  Path path() {
    final width = (end.dx - start.dx).abs();
    var path = Path();
    if (start.dy == end.dy) {
      path
        ..moveTo(start.dx, start.dy)
        ..lineTo(end.dx, end.dy);
    } else {
      double endAnchorX = width * (1 - bend);
      if (width - endAnchorX < 75) {
        endAnchorX = 75;
      }
      path
        ..moveTo(start.dx, start.dy)
        ..cubicTo(
          width * bend,
          start.dy,
          endAnchorX,
          end.dy,
          end.dx,
          end.dy,
        );
    }
    return path;
    return ArrowPath.make(path: path, isAdjusted: true);
  }
}

class ConnectionsPainter extends CustomPainter {
  final List<Connection> connections;

  const ConnectionsPainter({this.connections = const []});

  @override
  void paint(Canvas canvas, Size size) {
    final connectionPaint = Paint()
      ..color = Colors.black
      ..strokeWidth = 2
      ..style = PaintingStyle.stroke
      ..strokeJoin = StrokeJoin.round;

    for (final connection in connections) {
      final path = connection.path();
      canvas.drawPath(path, connectionPaint);
    }

    final tipPaint = Paint()
      ..color = Colors.black
      ..strokeWidth = 2
      ..style = PaintingStyle.stroke
      ..strokeJoin = StrokeJoin.round;

    for (final connection in connections) {
      final path =
          arrowTip(connection.end, connection.tipAngle, connection.tipLength);
      canvas.drawPath(path, tipPaint);
    }
  }

  @override
  bool shouldRepaint(ConnectionsPainter oldDelegate) => true;
}

Path arrowTip(Offset point, double angle, double length) {
  angle = _toRadian(180 - angle);
  final start = Offset(length * cos(angle), length * sin(angle));
  final end = Offset(length * cos(-angle), length * sin(-angle));
  return Path()
    ..moveTo(point.dx, point.dy)
    ..relativeLineTo(start.dx, start.dy)
    ..moveTo(point.dx, point.dy)
    ..relativeLineTo(end.dx, end.dy);
}

double _toRadian(double degree) => degree * pi / 180;


linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "connection_arrow");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "connection_arrow");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


lib/connection_arrow.dart
import 'dart:math';

import 'package:flutter/material.dart';

class Connection {
  final Offset start;
  final Offset end;
  final double bend;
  final Paint? stroke;
  final TipMaker? tip;

  const Connection(
      {required this.start,
      required this.end,
      this.bend = 0.5,
      this.stroke,
      this.tip});

  Path path() {
    final width = (end.dx - start.dx).abs();
    double endAnchorX = width * (1 - bend);
    if (width - endAnchorX < 75) {
      endAnchorX = 75;
    }
    return Path()
      ..moveTo(start.dx, start.dy)
      ..cubicTo(
        width * bend,
        start.dy,
        endAnchorX,
        end.dy,
        end.dx,
        end.dy,
      );
  }
}

class ConnectionsPainter extends CustomPainter {
  final List<Connection> connections;

  const ConnectionsPainter({this.connections = const []});

  @override
  void paint(Canvas canvas, Size size) {
    for (final connection in connections) {
      final paint = connection.stroke ?? defaultPaint;
      final path = connection.path();
      canvas.drawPath(path, paint);
    }

    for (final connection in connections) {
      connection.tip?.paint(canvas, connection.end);
    }
  }

  static final defaultPaint = Paint()
    ..color = Colors.black
    ..strokeWidth = 2
    ..style = PaintingStyle.stroke
    ..strokeJoin = StrokeJoin.round;

  @override
  bool shouldRepaint(ConnectionsPainter oldDelegate) => true;
}

abstract class TipMaker {
  Paint? get stroke;
  Paint? get fill;

  void paint(Canvas canvas, Offset point);
}

class ArrowTip implements TipMaker {
  final double angle;
  final double length;
  @override
  late final Paint? stroke;
  @override
  late final Paint? fill;

  ArrowTip({this.angle = 30, this.length = 15, Paint? stroke, Paint? fill}) {
    if (stroke == null && fill == null) {
      this.stroke = Paint()
        ..color = Colors.black
        ..strokeWidth = 2
        ..style = PaintingStyle.stroke
        ..strokeJoin = StrokeJoin.round;
      this.fill = null;
    } else {
      this.stroke = stroke;
      this.fill = fill;
    }
  }

  Path _path(Offset point) {
    double angle = _toRadian(180 - this.angle);
    final start = Offset(length * cos(angle), length * sin(angle));
    final end = Offset(length * cos(-angle), length * sin(-angle));
    return Path()
      ..moveTo(point.dx, point.dy)
      ..relativeLineTo(start.dx, start.dy)
      ..moveTo(point.dx, point.dy)
      ..relativeLineTo(end.dx, end.dy);
  }

  @override
  void paint(Canvas canvas, Offset point) {
    if (stroke != null) canvas.drawPath(_path(point), stroke!);
    if (fill != null) canvas.drawPath(_path(point), fill!);
  }
}

class CircleTip implements TipMaker {
  final double radius;
  @override
  late final Paint? stroke;
  @override
  late final Paint? fill;

  CircleTip({this.radius = 5, Paint? stroke, Paint? fill}) {
    if (stroke == null && fill == null) {
      this.stroke = Paint()
        ..color = Colors.black
        ..strokeWidth = 2
        ..style = PaintingStyle.stroke
        ..strokeJoin = StrokeJoin.round;
      this.fill = Paint()
        ..color = Colors.white
        ..style = PaintingStyle.fill;
    } else {
      this.stroke = stroke;
      this.fill = fill;
    }
  }

  @override
  void paint(Canvas canvas, Offset point) {
    if (stroke != null) canvas.drawCircle(point, radius, stroke!);
    if (fill != null) canvas.drawCircle(point, radius, fill!);
  }
}

double _toRadian(double degree) => degree * pi / 180;


lib/horizontal_arrow.dart
import 'package:flutter/material.dart';
import 'package:arrow_path/arrow_path.dart';

export 'arrow_painter.dart';

/*
class HorizontalArrow extends CustomPainter {
  final ArrowConfig config;

  const HorizontalArrow({required this.config});

  @override
  void paint(Canvas canvas, Size size) {
    final startPoint = config.startPoint(size);
    final startAnchor = config.startAnchor(size);
    final endPoint = config.endPoint(size);
    final endAnchor = config.endAnchor(size);
    Path bezierPath = Path()
      ..moveTo(startPoint.dx, startPoint.dy)
      ..cubicTo(
        startAnchor.dx,
        startAnchor.dy,
        endAnchor.dx,
        endAnchor.dy,
        endPoint.dx,
        endPoint.dy,
      );
    bezierPath = ArrowPath.make(path: bezierPath);

    final bezierPaint = Paint()
      ..color = Colors.black
      ..strokeWidth = 2
      ..style = PaintingStyle.stroke
      ..strokeJoin = StrokeJoin.round;

    canvas.drawPath(bezierPath, bezierPaint);
  }

  @override
  bool shouldRepaint(HorizontalArrow oldDelegate) =>
      oldDelegate.config.isEqual(config);
}

class ArrowConfig {
  final double bend;
  final CurveDirection direction;

  ArrowConfig({this.bend = 0.5, this.direction = CurveDirection.topToBottom});

  Offset startPoint(Size size) {
    switch(direction) {
      case CurveDirection.topToBottom:
        return Offset(0, 0);
      case CurveDirection.bottomToTop:
        return Offset(0, size.height);
      case CurveDirection.straight:
        return Offset(0, size.height / 2);
    }
  }
  Offset startAnchor(Size size) {
    Offset(size.width * bend, topToBottom ? 0 : size.height)
  }
  Offset endPoint(Size size) =>
      Offset(size.width, topToBottom ? size.height : 0);
  Offset endAnchor(Size size) =>
      Offset(size.width * (1 - bend), topToBottom ? size.height : 0);

  bool isEqual(ArrowConfig other) =>
      bend == other.bend && topToBottom == other.topToBottom;
}

enum CurveDirection {
  topToBottom,
  straight,
  bottomToTop,
}
 */

lib/curvy_connection.dart
import 'dart:math';

import 'package:flutter/material.dart';

class Connection {
  final Offset start;
  final Offset end;
  final double bend;
  final Paint? stroke;
  final TipMaker? tip;

  const Connection(
      {required this.start,
      required this.end,
      this.bend = 0.5,
      this.stroke,
      this.tip});

  bool isEqual(Connection other) =>
      start == other.start &&
      end == other.end &&
      bend == other.bend &&
      stroke == other.stroke &&
      tip == other.tip;

  Path path() {
    final width = (end.dx - start.dx).abs();
    double endAnchorX = width * (1 - bend);
    if (width - endAnchorX < 75) {
      endAnchorX = 75;
    }
    return Path()
      ..moveTo(start.dx, start.dy)
      ..cubicTo(
        start.dx + width * bend,
        start.dy,
        start.dx + endAnchorX,
        end.dy,
        end.dx,
        end.dy,
      );
  }
}

class ConnectionsPainter extends CustomPainter {
  final List<Connection> connections;

  const ConnectionsPainter({this.connections = const []});

  @override
  void paint(Canvas canvas, Size size) {
    for (final connection in connections) {
      final paint = connection.stroke ?? defaultPaint;
      final path = connection.path();
      canvas.drawPath(path, paint);
    }

    for (final connection in connections) {
      connection.tip?.paint(canvas, connection.end);
    }
  }

  static final defaultPaint = Paint()
    ..color = Colors.black
    ..strokeWidth = 2
    ..style = PaintingStyle.stroke
    ..strokeJoin = StrokeJoin.round;

  @override
  bool shouldRepaint(ConnectionsPainter oldDelegate) {
    if (connections.length != oldDelegate.connections.length) {
      return true;
    }
    for (int i = 0; i < connections.length; i++) {
      if (connections[i] != oldDelegate.connections[i]) {
        return true;
      }
    }
    return false;
  }
}

abstract class TipMaker {
  Paint? get stroke;
  Paint? get fill;

  void paint(Canvas canvas, Offset point);
}

class ArrowTip implements TipMaker {
  final double angle;
  final double length;
  @override
  late final Paint? stroke;
  @override
  late final Paint? fill;

  ArrowTip({this.angle = 30, this.length = 15, Paint? stroke, Paint? fill}) {
    if (stroke == null && fill == null) {
      this.stroke = Paint()
        ..color = Colors.black
        ..strokeWidth = 2
        ..style = PaintingStyle.stroke
        ..strokeJoin = StrokeJoin.round;
      this.fill = null;
    } else {
      this.stroke = stroke;
      this.fill = fill;
    }
  }

  Path _path(Offset point) {
    double angle = _toRadian(180 - this.angle);
    final start = Offset(length * cos(angle), length * sin(angle));
    final end = Offset(length * cos(-angle), length * sin(-angle));
    return Path()
      ..moveTo(point.dx, point.dy)
      ..relativeLineTo(start.dx, start.dy)
      ..moveTo(point.dx, point.dy)
      ..relativeLineTo(end.dx, end.dy);
  }

  @override
  void paint(Canvas canvas, Offset point) {
    if (stroke != null) canvas.drawPath(_path(point), stroke!);
    if (fill != null) canvas.drawPath(_path(point), fill!);
  }
}

class CircleTip implements TipMaker {
  final double radius;
  @override
  late final Paint? stroke;
  @override
  late final Paint? fill;

  CircleTip({this.radius = 5, Paint? stroke, Paint? fill}) {
    if (stroke == null && fill == null) {
      this.stroke = Paint()
        ..color = Colors.black
        ..strokeWidth = 2
        ..style = PaintingStyle.stroke
        ..strokeJoin = StrokeJoin.round;
      this.fill = Paint()
        ..color = Colors.white
        ..style = PaintingStyle.fill;
    } else {
      this.stroke = stroke;
      this.fill = fill;
    }
  }

  @override
  void paint(Canvas canvas, Offset point) {
    if (stroke != null) canvas.drawCircle(point, radius, stroke!);
    if (fill != null) canvas.drawCircle(point, radius, fill!);
  }
}

double _toRadian(double degree) => degree * pi / 180;
